Add support for ignoring deprecations by path

markstory opened the pull request:
Being able to have deprecations on globally but silence problematic
parts of your application or a noisy plugin that can't be upgraded just
yet allows developers more control when upgrading.

Instead of having to hope they don't add new deprecations they can fix
a part of their application and then remove it from the deprecation
ignore patterns.

I've used glob here instead of regex as I don't think the power of regex
is needed for this.

Refs cakephp/cakephp#14915
Refs cakephp/app#803

markstory committed:
Add support for ignoring deprecations by path

Being able to have deprecations on globally but silence problematic
parts of your application or a noisy plugin that can't be upgraded just
yet allows developers more control when upgrading.

Instead of having to hope they don't add new deprecations they can fix
a part of their application and then remove it from the deprecation
ignore patterns.

I've used glob here instead of regex as I don't think the power of regex
is needed for this.

Refs cakephp/cakephp#14915

markstory committed:
Remove debug code.

markstory committed:
Update wording.

garas commented on src/Core/functions.php:
```diff
@@ -292,10 +292,20 @@ function deprecationWarning(string $message, int $stackFrame = 1): void
                 '%s - %s, line: %s' . "\n" .
                 ' You can disable deprecation warnings by setting `Error.errorLevel` to' .
-                ' `E_ALL & ~E_USER_DEPRECATED` in your config/app.php.',
+                " `E_ALL & ~E_USER_DEPRECATED`, or add `{$relative}` to " .
```
Add `str_replace(DIRECTORY_SEPARATOR, '/', $relative)` through `sprintf` placeholder.

markstory committed:
Normalize paths better.

markstory committed:
Fix phpcs error.

markstory committed:
Normalize to / instead of DIRECTORY_SEPARATOR

I think \ might be causing problems with fnmatch() on windows as builds
have failed a few times.

markstory committed:
Fix configuration name to be clearer.

garas commented on src/Core/functions.php:
```diff
@@ -292,13 +292,24 @@ function deprecationWarning(string $message, int $stackFrame = 1): void
             $message = sprintf(
                 '%s - %s, line: %s' . "\n" .
                 ' You can disable deprecation warnings by setting `Error.errorLevel` to' .
-                ' `E_ALL & ~E_USER_DEPRECATED` in your config/app.php.',
+                ' `E_ALL & ~E_USER_DEPRECATED`, or add `%s` to ' .
+                ' `Error.ignoredDeprecationPaths` in your `config/app.php` to mute deprecations.',
```
"You can disable **all** deprecation warnings by  ...., or add .... to mute deprecations **in this file only**."

markstory committed:
Improve error message wording.
